src/v/cloud_storage/materialized_segments.cc

```diff
@@ -313,19 +313,8 @@ void materialized_segments::maybe_trim_segment(
         // this will delete and unlink the object from
         // _materialized collection
         if (st.parent) {
```

Comment:
> Removes redudndant checks in materialized_segments, now when
remote_partition stops, it unlinks the materialized segments from other
containers first, at which point it is the only one that can stop them.
Caveat (explained in docstrings): materialized_segments runs
trim_segments on a timer. If this method has started, it will run over
all segments and mark the ones which are too old for eviction. As this
is a synchronous function with no yield points, it will run from start
to finish without interruption.
The remote_partition::stop on the other hand stops segments directly, is
async and can yield at multiple points. So there is a possibility that
remote_partition::stop yields and trim_segments runs. But because we
clear out the segments before the first yield point, once
remote_partition::stop yields, materialized_segments will not have any
segments to stop, thus making the operation safe.
However if in future materialized_segments::trim_segments becomes async
and yields while processing segments, and _then_ remote_partition::stop
clears out the segments in its map, this can lead to a crash when
trim_segments resumes. So trim_segments must never be async.
An alternative would be to have a mutex in remote_partition which is
required before running stop OR trim_segments.

Thank you for writing thorough commit messages :heart: